tomviz/SaveLoadTemplateReaction.cxx

```diff
@@ -0,0 +1,121 @@
+/* This source file is part of the Tomviz project, https://tomviz.org/.
+   It is released under the 3-Clause BSD License, see "LICENSE". */
+
+#include "SaveLoadTemplateReaction.h"
+
+#include "ActiveObjects.h"
+#include "ModuleManager.h"
+#include "Pipeline.h"
+#include "RecentFilesMenu.h"
+#include "Utilities.h"
+
+#include <QApplication>
+#include <QDebug>
+#include <QInputDialog>
+#include <QJsonArray>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <QString>
+
+namespace tomviz {
+
+SaveLoadTemplateReaction::SaveLoadTemplateReaction(QAction* parentObject, bool load, QString filename)
+  : pqReaction(parentObject), m_load(load), m_filename(filename)
+{}
+
+void SaveLoadTemplateReaction::onTriggered()
+{
+  if (m_load) {
+    loadTemplate(m_filename);
+  } else {
+    saveTemplate();
+  }
+}
+
+bool SaveLoadTemplateReaction::saveTemplate()
+{
+  bool ok;
+  QString text = QInputDialog::getText(tomviz::mainWidget(), tr("Save Pipeline Template"),
+                                         tr("Template Name:"), QLineEdit::Normal, QString(), &ok);
+  QString fileName = text.replace(" ", "_");
+  if (ok && !text.isEmpty()) {
+    QString path = QApplication::applicationDirPath() +
+                    "/../share/tomviz/templates/" + fileName + ".tvsm";
+    return SaveLoadTemplateReaction::saveTemplate(path);
+  }
+  return false;
+}
+
+bool SaveLoadTemplateReaction::loadTemplate(const QString& fileName)
+{
+  QString path = QApplication::applicationDirPath() +
+                 "/../share/tomviz/templates/" + fileName + ".tvsm";
+  QFile openFile(path);
+  openFile.open(QIODevice::ReadOnly);
+  QJsonParseError error;
+  auto contents = openFile.readAll();
+  auto doc = QJsonDocument::fromJson(contents, &error);
+
+  // Get the parent data source, as well as the active (i.e. data and output)
+  auto activeParent = ActiveObjects::instance().activeParentDataSource();
+  auto activeData = ActiveObjects::instance().activeDataSource();
+  ModuleManager::instance().removeAllModules(activeData);
```

Comment:
Why do we remove all modules? It seems that if we don't touch any existing operators we shouldn't touch existing modules?

Reply:
I think my initial thought was that if someone is loading a template that includes modules they might not want the default modules as well, but you're right - if it's being assumed that any existing operators should be left in place, the same should be done with modules to keep the behavior consistent.

Reply:
@cryos Do you have any thoughts/feels on what this behavior should be?

Reply:
I think templates should be expected to be able to combine, i.e. I would add one template to another possibly. I wonder if we might actually add a clear all to the context menu, but I would rather default to appending templates to the end of a pipeline, allowing users to freely mix their starting point with canned templates.